python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -120,6 +128,218 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
+struct ScalarUdfAggregator : public compute::KernelState {
+  virtual Status Consume(compute::KernelContext* ctx, const compute::ExecSpan& batch) = 0;
+  virtual Status MergeFrom(compute::KernelContext* ctx, compute::KernelState&& src) = 0;
+  virtual Status Finalize(compute::KernelContext* ctx, Datum* out) = 0;
+};
+
+arrow::Status AggregateUdfConsume(compute::KernelContext* ctx, const compute::ExecSpan& batch) {
+  return checked_cast<ScalarUdfAggregator*>(ctx->state())->Consume(ctx, batch);
+}
+
+arrow::Status AggregateUdfMerge(compute::KernelContext* ctx, compute::KernelState&& src,
+                                compute::KernelState* dst) {
+  return checked_cast<ScalarUdfAggregator*>(dst)->MergeFrom(ctx, std::move(src));
+}
+
+arrow::Status AggregateUdfFinalize(compute::KernelContext* ctx, arrow::Datum* out) {
+  return checked_cast<ScalarUdfAggregator*>(ctx->state())->Finalize(ctx, out);
+}
+
+ScalarAggregateUdfContext::~ScalarAggregateUdfContext() {
+  if (_Py_IsFinalizing()) {
+      Py_DECREF(this->state);
+    }
+}
+
+struct PythonScalarUdfAggregatorImpl : public ScalarUdfAggregator {
+
+  ScalarAggregateInitUdfWrapperCallback init_cb;
+  ScalarAggregateConsumeUdfWrapperCallback consume_cb;
+  ScalarAggregateMergeUdfWrapperCallback merge_cb;
+  ScalarAggregateFinalizeUdfWrapperCallback finalize_cb;
+  std::shared_ptr<OwnedRefNoGIL> init_function;
+  std::shared_ptr<OwnedRefNoGIL> consume_function;
+  std::shared_ptr<OwnedRefNoGIL> merge_function;
+  std::shared_ptr<OwnedRefNoGIL> finalize_function;
+  std::shared_ptr<DataType> output_type;
+
+
+  PythonScalarUdfAggregatorImpl(ScalarAggregateInitUdfWrapperCallback init_cb,
+   ScalarAggregateConsumeUdfWrapperCallback consume_cb,
+   ScalarAggregateMergeUdfWrapperCallback merge_cb,
+   ScalarAggregateFinalizeUdfWrapperCallback finalize_cb,
+   std::shared_ptr<OwnedRefNoGIL> init_function,
+   std::shared_ptr<OwnedRefNoGIL> consume_function,
+   std::shared_ptr<OwnedRefNoGIL> merge_function,
+   std::shared_ptr<OwnedRefNoGIL> finalize_function,
+            const std::shared_ptr<DataType>& output_type) : init_cb(init_cb),
+            consume_cb(consume_cb),
+            merge_cb(merge_cb),
+            finalize_cb(finalize_cb),
+            init_function(init_function),
+            consume_function(consume_function),
+            merge_function(merge_function),
+            finalize_function(finalize_function),
+            output_type(output_type) {
+              Init(init_cb, init_function);
+            }
+
+  ~PythonScalarUdfAggregatorImpl() {
+    if (_Py_IsFinalizing()) {
+      init_function->detach();
+      consume_function->detach();
+      merge_function->detach();
+      finalize_function->detach();
+    }
+  }
+
+  void Init(ScalarAggregateInitUdfWrapperCallback& init_cb , std::shared_ptr<OwnedRefNoGIL>& init_function) {
+    auto st =  SafeCallIntoPython([&]() -> Status { 
+      OwnedRef result(init_cb(init_function->obj()));
+      PyObject* init_res = result.obj();
```

Comment:
Probably need to do sanity checking (e.g. not null, etc.) on things returned from user callbacks.